README.md
# Soccer Predictor

[![Latest npm version](https://img.shields.io/npm/v/soccer-predictor.svg)](https://www.npmjs.com/package/soccer-predictor)
[![Build Status](https://img.shields.io/travis/CookPete/soccer-predictor/master.svg)](https://travis-ci.org/CookPete/soccer-predictor)
[![Dependency Status](https://img.shields.io/david/CookPete/soccer-predictor.svg)](https://david-dm.org/CookPete/soccer-predictor)
[![devDependency Status](https://img.shields.io/david/dev/CookPete/soccer-predictor.svg)](https://david-dm.org/CookPete/soccer-predictor?type=dev)
[![Test Coverage](https://img.shields.io/codecov/c/github/cookpete/soccer-predictor.svg)](https://codecov.io/gh/CookPete/soccer-predictor)

A JS library that predicts soccer match outcomes using basic mathematics.

## Installation

Use [yarn](https://yarnpkg.com) or [npm](https://www.npmjs.com) to install.

```shell
yarn add soccer-predictor           # yarn
npm install soccer-predictor --save # npm
```

### Install locally

```shell
git clone https://github.com/CookPete/soccer-predictor.git
cd soccer-predictor
yarn # or npm install
npm run build
```

### Testing locally

```shell
npm test # Run tests
```

## Usage

```js
import { analyseResults, calculateProbabilities } from 'soccer-predictor'

// Some basic example data
const results = [
  { homeTeam: 'Team A', awayTeam: 'Team B', homeGoals: 1, awayGoals: 0 },
  { homeTeam: 'Team B', awayTeam: 'Team C', homeGoals: 2, awayGoals: 1 },
  { homeTeam: 'Team C', awayTeam: 'Team A', homeGoals: 1, awayGoals: 3 }
]

// Function to map match data to soccer-predictor data
function getMatchDetails (match) {
  return {
    homeTeamName: match.homeTeam,
    awayTeamName: match.awayTeam,
    homeGoals: match.homeGoals,
    awayGoals: match.awayGoals
  }
}

// Parse results into an array of teams with calculated stats
const teams = analyseResults(results, getMatchDetails)

// Use calculateProbabilities to calculate the chance of
// various outcomes of a match between two teams
const probabilities = calculateProbabilities(teams[0], teams[1])

console.log(probabilities.result) // Probability of a home win, away win or draw
// {
//   home: 0.657541673613264,
//   draw: 0.252533180170396,
//   away: 0.089924594462736
// }

console.log(probabilities.scores[1][0]) // Probability of a 1-0 result
// 0.2300324502673927

console.log(probabilities.over['2.5']) // Probability of over 2.5 goals
// 0.2895346409101279

console.log(probabilities.btts.yes) // Probability of both teams to score
// 0.2429353553660894

```

For more examples, see [`examples`](./examples).

## FAQ

### How does it work?

There are plenty of articles available that go into depth about basic poisson prediction. Essentially the library parses a series of match results and provides functions to return a rough percentage likelyhood of various outcomes of a match between two teams.

Further reading:

* https://plus.maths.org/content/understanding-uncertainty-football-crazy
* http://opisthokonta.net/?p=296
* https://www.sbo.net/strategy/football-prediction-model-poisson-distribution/

### Isn’t it called football?

Yes, but `football-predictor` seems too ambigious. Everyone knows what [soccer](https://en.wikipedia.org/wiki/Soccer) is.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This project adheres to [Semantic Versioning](http://semver.org/).

Generated by [auto-changelog](https://github.com/CookPete/auto-changelog)


#### [v0.0.2](https://github.com/CookPete/soccer-predictor/compare/v0.0.1...v0.0.2)
> 27 September 2017

* Add code coverage [`fb1ca95`](https://github.com/CookPete/soccer-predictor/commit/fb1ca958b4d55df7254e9c5f71df0451630db337)
* Add auto-changelog [`66a4762`](https://github.com/CookPete/soccer-predictor/commit/66a476211d726d8d7605d3a555dc8b48edf35dc7)
* Add repo url to package.json [`48fab0d`](https://github.com/CookPete/soccer-predictor/commit/48fab0d17d0b987b51e8239645e01a9d1411b9ba)


#### v0.0.1
> 27 April 2017

* Add yarn lockfile [`2d7e489`](https://github.com/CookPete/soccer-predictor/commit/2d7e489ba1b46d7de492353729d81588f9cdcd2a)
* Add basic prediction functions [`80650cd`](https://github.com/CookPete/soccer-predictor/commit/80650cdbfeb3cfba82a3e0ceec90bc8de6e988d6)
* Bump packages [`22d2399`](https://github.com/CookPete/soccer-predictor/commit/22d23990ba8287f43d2718ad398cd2e7ad5b3272)


test/math.js
import { describe, it } from 'mocha'
import { expect } from 'chai'

import { average, poisson } from '../src/utils/math'

// http://faculty.wwu.edu/auer/DSCI205/bbs-e12-POIS.pdf
const POISSON_TABLE = [
  [0.9048, 0.8187, 0.7408, 0.6703, 0.6065, 0.5488, 0.4966, 0.4493, 0.4066, 0.3679],
  [0.0905, 0.1637, 0.2222, 0.2681, 0.3033, 0.3293, 0.3476, 0.3595, 0.3659, 0.3679],
  [0.0045, 0.0164, 0.0333, 0.0536, 0.0758, 0.0988, 0.1217, 0.1438, 0.1647, 0.1839],
  [0.0002, 0.0011, 0.0033, 0.0072, 0.0126, 0.0198, 0.0284, 0.0383, 0.0494, 0.0613],
  [0.0000, 0.0001, 0.0003, 0.0007, 0.0016, 0.0030, 0.0050, 0.0077, 0.0111, 0.0153],
  [0.0000, 0.0000, 0.0000, 0.0001, 0.0002, 0.0004, 0.0007, 0.0012, 0.0020, 0.0031]
]

describe('average', () => {
  it('returns the average of an array', () => {
    expect(average([1, 2, 3])).to.equal(2)
  })
})

describe('poisson', () => {
  it('returns correct values', () => {
    for (let x = 0; x !== 6; x++) {
      for (let y = 0; y !== 10; y++) {
        const l = (y + 1) / 10
        expect(parseFloat(poisson(x, l).toFixed(4))).to.equal(POISSON_TABLE[x][y])
      }
    }
  })
})


test/index.js
import { describe, it } from 'mocha'
import { expect } from 'chai'

import { add } from '../src/index'

describe('add', () => {
  it('returns correct value', () => {
    expect(add(1, 2)).to.equal(3)
  })
})


test/parse.js
import { describe, it } from 'mocha'
import { expect } from 'chai'

import { parseCSV } from '../src/utils/parse'

describe('parseCSV', () => {
  const expected = [
    { Name: 'John', Age: 40, Gender: 'Male' },
    { Name: 'Barry', Age: 35, Gender: 'Male' },
    { Name: 'Mary', Age: 32, Gender: 'Female' }
  ]

  it('parses a CSV string', () => {
    const input = `Name,Age,Gender\nJohn,40,Male\nBarry,35,Male\nMary,32,Female`
    expect(parseCSV(input)).to.deep.equal(expected)
  })

  it('skips blank lines', () => {
    const input = `Name,Age,Gender\n\n\n\nJohn,40,Male\n\nBarry,35,Male\n\nMary,32,Female\n\n`
    expect(parseCSV(input)).to.deep.equal(expected)
  })

  it('skips comma lines', () => {
    const input = `Name,Age,Gender\n\n\n\nJohn,40,Male\n\nBarry,35,Male\n\nMary,32,Female\n\n,,\n\n`
    expect(parseCSV(input)).to.deep.equal(expected)
  })

  it('takes a delimiter parameter', () => {
    const input = `Name\tAge\tGender\nJohn\t40\tMale\nBarry\t35\tMale\nMary\t32\tFemale`
    expect(parseCSV(input, '\t')).to.deep.equal(expected)
  })
})


src/teams.js
import { average } from './utils/math'
import { iterate } from './utils/object'

const MIN_SCORE = 0.1

// Takes an array of results and a function to return match details
// and returns an array of teams with calculated performance stats
export function analyseResults (results, getMatchDetails) {
  if (!results || results instanceof Array === false) {
    throw new Error('First parameter must be an array of results')
  }

  if (typeof getMatchDetails !== 'function') {
    throw new Error('Second parameter must be a function')
  }

  let teams = {}

  for (let result of results) {
    const { homeTeamName, awayTeamName, homeGoals, awayGoals } = getMatchDetails(result)

    const homeTeam = teams[homeTeamName] || newTeam()
    const awayTeam = teams[awayTeamName] || newTeam()

    homeTeam.played++
    homeTeam.home.played++
    homeTeam.goals.for += homeGoals
    homeTeam.goals.against += awayGoals
    homeTeam.home.goals.for += homeGoals
    homeTeam.home.goals.against += awayGoals

    awayTeam.played++
    awayTeam.away.played++
    awayTeam.goals.for += awayGoals
    awayTeam.goals.against += homeGoals
    awayTeam.away.goals.for += awayGoals
    awayTeam.away.goals.against += homeGoals

    if (homeGoals > awayGoals) {
      homeTeam.wins++
      homeTeam.home.wins++
      homeTeam.form.push('W')
      homeTeam.home.form.push('W')
      awayTeam.losses++
      awayTeam.away.losses++
      awayTeam.form.push('L')
      awayTeam.away.form.push('L')
    }

    if (homeGoals === awayGoals) {
      homeTeam.draws++
      homeTeam.home.draws++
      homeTeam.form.push('D')
      homeTeam.home.form.push('D')
      awayTeam.draws++
      awayTeam.away.draws++
      awayTeam.form.push('D')
      awayTeam.away.form.push('D')
    }

    if (homeGoals < awayGoals) {
      homeTeam.losses++
      homeTeam.home.losses++
      homeTeam.form.push('L')
      homeTeam.home.form.push('L')
      awayTeam.wins++
      awayTeam.away.wins++
      awayTeam.form.push('W')
      awayTeam.away.form.push('W')
    }

    teams[homeTeamName] = homeTeam
    teams[awayTeamName] = awayTeam
  }

  const averages = getAverages(results, getMatchDetails, teams)

  return Object.keys(teams).map(name => {
    return {
      ...teams[name],
      name,
      stats: getTeamStats(teams[name], averages)
    }
  })
}

function newTeam () {
  return {
    played: 0,
    wins: 0,
    draws: 0,
    losses: 0,
    goals: { for: 0, against: 0 },
    form: [],
    home: { played: 0, wins: 0, draws: 0, losses: 0, goals: { for: 0, against: 0 }, form: [] },
    away: { played: 0, wins: 0, draws: 0, losses: 0, goals: { for: 0, against: 0 }, form: [] }
  }
}

function getAverages (results, getMatchDetails, teams) {
  // Arrays of goal counts to be averaged out
  let averages = {
    for: [],
    against: [],
    home: { for: [], against: [] },
    away: { for: [], against: [] },
    expectedGoals: { home: [], away: [] } // Same for all teams
  }

  for (let match of results) {
    const { homeGoals, awayGoals } = getMatchDetails(match)
    averages.expectedGoals.home.push(homeGoals)
    averages.expectedGoals.away.push(awayGoals)
  }

  for (let key of Object.keys(teams)) {
    const team = teams[key]
    averages.for.push(team.goals.for)
    averages.against.push(team.goals.against)
    averages.home.for.push(team.home.goals.for)
    averages.home.against.push(team.home.goals.against)
    averages.away.for.push(team.away.goals.for)
    averages.away.against.push(team.away.goals.against)
  }

  return iterate(averages, average)
}

// as: Attacking strength
// dw: Defense weakness
function getTeamStats (team, averages) {
  return {
    as: getScore(team.goals.for / averages.for),
    dw: getScore(team.goals.against / averages.against),
    home: {
      as: getScore(team.home.goals.for / averages.home.for),
      dw: getScore(team.home.goals.against / averages.home.against)
    },
    away: {
      as: getScore(team.away.goals.for / averages.away.for),
      dw: getScore(team.away.goals.against / averages.away.against)
    },
    expectedGoals: averages.expectedGoals // Same for all teams
  }
}

function getScore (score) {
  return Math.max(score, MIN_SCORE)
}


examples/parse.js
import { readFile } from 'fs'
import { join } from 'path'

import { parseCSV, analyseResults, calculateProbabilities } from '../src/index'

const file = join(__dirname, 'data', '2015-prem.csv')

readFile(file, 'utf8', (err, data) => {
  if (err) throw err

  // Use parseCSV to get an array of results from the CSV file
  const results = parseCSV(data)
  console.log(`${results.length} results from parsing CSV file`)

  // Use analyseResults to return an array of teams with stats
  const teams = analyseResults(results, getMatchDetails)
  console.log(`${teams.length} teams from analysing results`)

  // Now lets calculate probabilities of the first and
  // second teams in the array as an example
  const homeTeam = teams[0]
  const awayTeam = teams[1]
  const probs = calculateProbabilities(homeTeam, awayTeam)

  console.log(`Prediction for ${homeTeam.name} vs ${awayTeam.name}:`)
  console.log(`${percent(probs.result.home)} home win`)
  console.log(`${percent(probs.result.draw)} draw`)
  console.log(`${percent(probs.result.away)} away win`)
})

// Function to return teams and scores based on the
// CSV file headings, to be used by analyseResults
function getMatchDetails (data) {
  return {
    homeTeamName: data.HomeTeam,
    awayTeamName: data.AwayTeam,
    homeGoals: data.FTHG,
    awayGoals: data.FTAG
  }
}

function percent (percent) {
  return (percent * 100).toFixed(1) + '%'
}


src/prediction.js
import { poisson } from './utils/math'

const MAX_GOALS = 10 // Number of goals to scorecast

// Takes two teams from the result of analyseResults
// and calculates probability of different match outcomes
// using a basic poisson formula
// https://plus.maths.org/content/understanding-uncertainty-football-crazy
export function calculateProbabilities (homeTeam, awayTeam) {
  // Expected number of goals for teams when home and away
  let homeExpectedGoals = homeTeam.stats.expectedGoals.home
  let awayExpectedGoals = awayTeam.stats.expectedGoals.away

  // Calculate performance multipliers based on team scores
  const homeMultiplier = homeTeam.stats.as * awayTeam.stats.dw
  const awayMultiplier = awayTeam.stats.as * homeTeam.stats.dw

  // Apply multipliers
  homeExpectedGoals *= homeMultiplier
  awayExpectedGoals *= awayMultiplier

  // Predict an outcome based on expected goals scored
  return scorecast(homeExpectedGoals, awayExpectedGoals)
}

function scorecast (homeExpectedGoals, awayExpectedGoals) {
  const result = { home: 0, away: 0, draw: 0 }
  const scores = []
  const under = {
    '0.5': 0,
    '1.5': 0,
    '2.5': 0,
    '3.5': 0,
    '4.5': 0
  }
  const over = {
    '0.5': 0,
    '1.5': 0,
    '2.5': 0,
    '3.5': 0,
    '4.5': 0
  }
  const btts = { yes: 0, no: 0 }

  for (let h = 0; h <= MAX_GOALS; h++) {
    scores[h] = []
    for (let a = 0; a <= MAX_GOALS; a++) {
      const probability = poisson(h, homeExpectedGoals) * poisson(a, awayExpectedGoals)

      // Home/away win
      if (h > a) result.home += probability
      if (h === a) result.draw += probability
      if (h < a) result.away += probability

      // Scores
      scores[h][a] = probability

      // BTTS
      if (h === 0 || a === 0) {
        btts.no += probability
      } else {
        btts.yes += probability
      }

      // Under/over
      for (let i = 0.5; i < 5; i++) {
        if (h + a < i) under[i] += probability
        if (h + a > i) over[i] += probability
      }
    }
  }

  return {
    result,
    scores,
    under,
    over,
    btts
  }
}


examples/basic.js
import { analyseResults, calculateProbabilities } from '../src'

// Some basic example data
const results = [
  { homeTeam: 'Team A', awayTeam: 'Team B', homeGoals: 1, awayGoals: 0 },
  { homeTeam: 'Team B', awayTeam: 'Team C', homeGoals: 2, awayGoals: 1 },
  { homeTeam: 'Team C', awayTeam: 'Team A', homeGoals: 1, awayGoals: 3 }
]

// Function to map match data to soccer-predictor data
function getMatchDetails (match) {
  return {
    homeTeamName: match.homeTeam,
    awayTeamName: match.awayTeam,
    homeGoals: match.homeGoals,
    awayGoals: match.awayGoals
  }
}

// Parse results into an array of teams with calculated stats
const teams = analyseResults(results, getMatchDetails)

// Use calculateProbabilities to calculate the chance of
// various outcomes of a match between two teams
const probabilities = calculateProbabilities(teams[0], teams[1])

console.log(probabilities.result) // Probability of a home win, away win or draw
// {
//   home: 0.657541673613264,
//   draw: 0.252533180170396,
//   away: 0.089924594462736
// }

console.log(probabilities.scores[1][0]) // Probability of a 1-0 result
// 0.2300324502673927

console.log(probabilities.over['2.5']) // Probability of over 2.5 goals
// 0.2895346409101279

console.log(probabilities.btts.yes) // Probability of both teams to score
// 0.2429353553660894
